Extracts ValueUpdater from ExtraBaggageFields for use in decoding (#1187)
openzipkin/brave: brave/RATIONALE.md

#### Internal state is an `Object[]`
`BaggageField` values are `String`, so to hold state means an association of
some kind between `BaggageField` and a single `String` value. Updates to
baggage are generally infrequent, sometimes never, and many times only when
extracting headers from a request. As such, we choose pair-wise layout in an
object array (`Object[]`).

To remind, pair-wise means an association `{k1=v1, k2=v2}` is laid out in an
array as `[k1, v1, k2, v2]`. The size of the association is `array.length/2`.
Retrieving a value means finding the key's index and returning `array[i+1]`

In our specific case, we add `null` values for all known fields. For example,
if `BaggageConfiguration` is initialized with fields countryCode and deviceId,
the initial state array is `[countryCode, null, deviceId, null]`

Retaining fields even when `null` allows among many things, the following:
* Queries for all potential field names (remote propagation)
* Constant time index lookup of predefined fields.

In practice, we use a constant `HashMap` to find the index of a predefined
fields. This is how we achieve constant time index lookup, while also keeping

> iteration performance of an array!

Author: this was just moved

### `Map` views over `BaggageField` values
As `Map` is a standard type, it is a safer type to expose to consumers of all
baggage fields. We decided to use an unmodifiable `Map` instead of an internal
type for use cases such as coalescing all baggage into a single header. This
map must also hide local fields from propagation.

#### Why not a standard `Map` type
We considered copying the internal state array to an existing `Map` type, such
as `LinkedHashMap`. However, doing so would add overhead regardless of it that
`Map` was ever used, or if that map had a value for the field the consumer was
interested in! Concretely, we'd pay to create the map, to copy the values in
the array into it, and also pay implicit cost of iterator allocation for
operations such as `entrySet()` even if the entry desired was not present. This
holds present even extending `AbstractMap`, as most operations, even `get()`
allocate iterators.

#### `UnsafeArrayMap`
We created `UnsafeArrayMap` as a view over our pair-wise internal state array.
The implementation is very simple except that we have a redaction use case to
address, which implies filtering keys. To address that, we keep a bitmap of all

> filtered keys and consider that when performing any scan operations.

Author: This also was just moved (I moved them as I noticed we already had a baggage heading!

In practice, we use a constant `HashMap` to find the index of a predefined
fields. This is how we achieve constant time index lookup, while also keeping
iteration performance of an array!

### `BaggageField.ValueUpdater`
`BaggageField.ValueUpdater` is a functional interface that allows field
updates, possibly to `null`.

It returns a `boolean` to inform callers an update was ignored, which can
happen when the value didn't change or the field is disallowed. We use this to
avoid unnecessary overhead when flushing changes to `MDC`.

Being a functional interface, it is also safer to use when exposing state to
outside callers. For example, we can provide a write view over a mutable array,
and decouple that connection when decoding completes.

### Encoding and decoding baggage from remote fields

`BaggageCodec` handles encoding and decoding of baggage fields for remote
propagation. There are some subtle points easy to miss about the design.

* Decode is called before a trace context exists, hence no `context` parameter.
  * First successful decode returns `true`.
    * This allows us to prioritize and leniently read, multiple header names.

>   * The `ValueUpdater parameter allows us to safely collect values from

Author: ```suggestion
  * The `ValueUpdater` parameter allows us to safely collect values from
```

    efficient than map).
    * The `context` parameter also allows secondary sampling to write down the
      `spanId`
  * The `request` parameter allows secondary sampling to use `SamplerFunction`
    on outbound requests.

#### Why not decode into a Map?
When looking at the encode vs decode side, it might seem curious why they don't
both implement `Map`. `Map` is helpful on the encode side. For example,
implementing  single-header encoding, is easier this way.

```java
joiner = Joiner.on(',').withKeyValueSeparator('=');

String encoded = joiner.join(baggageValues);
```

On the read side, the above iteration is stable because the underlying state is
immutable. There's no chance of a concurrent modification or otherwise
corrupting the encoded value. This applies even when the implementation ignores
the documentation and holds a reference to the map after encoding.

The reverse isn't true. Decoding has different concerns entirely.
* `BaggageField` instances are typically constant, so updates should be a
  function of `BaggageField`, not String.
* There may be multiple decoders contributing to the same baggage state. It is

>   more safe underneath to not expose read-ops, such as iterators.

Author: ```suggestion
  more safe underneath to prevent read-ops, such as iterators.
```

It is true that there are tools that can split common data structures into a
map. To do the same with another function could seem annoying.

Ex.
```java
splitter = Splitter.on(',').withKeyValueSeparator('=');

values = splitter.split(encoded);
```

However, if you look above, this assumes the key type is `String`. Such an
implementation would still need to loop over the result, possibly reuse an
existing `BaggageField`, then .. copy into the correct map!

Ex.
```java
splitter = Splitter.on(',').withKeyValueSeparator('=');

...
for (Entry<String, String> entry: splitter.split(encoded).entrySet()) {
  BaggageField field = lookupOrCreateField(entry.getKey());
  baggageValues.put(field, entry.getValue());
}
```

When put in this view, it isn't much different than just using `ValueUpdater`.
Ex.
```java
splitter = Splitter.on(',').withKeyValueSeparator('=');

...
for (Entry<String, String> entry: splitter.split(encoded).entrySet()) {
  BaggageField field = lookupOrCreateField(entry.getKey());
  valueUpdater.updateValue(field, entry.getValue());
}
```

A closing point is that the above splitter example is contrived, mainly to show
a design weakness in not choosing map.

> In practice, encoding formats typically have rules that splitters libraries

Author: ```suggestion
In practice, encoding formats typically have rules that splitter libraries
```

weren't written for. These include limits, character set constraints, etc.
Decoding a format is always harder than encoding due to rules about how to

> handle bad data. In general, it is better to assume a format-specific encoder

Author: ```suggestion
handle bad data. In general, it is better to assume a format-specific decoder
```
